Add, Update & Delete with Batch Support Working.

tushardhadiwal opened the pull request:
@jessebenson @jdhuntington @andrewjamconrad
I have implemented Back end with Add, Update & Delete functionality.
Partition ID Support is added.

Commit:
First step.

Commit:
Hlad PartitionID Support Added, BAsic Deelet Functionality Working.

jessebenson committed:
Whitespace.

jessebenson committed:
Support for PartitionId/Key/Value in queries and metadata.

jessebenson committed:
Support partition-based or cross-partition deletes.

Commit:
Basic Add Functionality Working

Commit:
Using TryAddSync with PartitionID Support for ID. Can add records to specific Partition & Adds to first available partition when ID not mentioned.

Commit:
"Update" Functionality Working with SetAsync as backbone.
Updates Value across all partitions if no PID Provides, if PID provided, Updates specific value for key in that partition.
Pending:Errorcodes handling

Commit:
Error Code Exceptions with HTTP Status Response handled for repetative add for existing key.

Commit:
Basic Batch Delete Working.

Commit:
Batch Upadte & Batch Add Support.

Commit:
Batch Update & Batch Add Support.

jessebenson commented:
@tushardhadiwal

jessebenson commented:
@jdhuntington @andrewjamconrad - code review!

jessebenson commented on samples/Basic/Basic.Common/Product.cs:
```diff
@@ -4,6 +4,6 @@ public sealed class Product
-		public int Quantity { get; set; }
-	}
+        public int Quantity { get; set; }
+    }
```
(nit):  whitespace.

Commit:
whitespace!

jessebenson commented on samples/Basic/Basic.ProductSvc/ProductSvc.cs:
```diff
@@ -14,7 +14,7 @@ namespace Basic.ProductSvc
 	/// <summary>
 	/// An instance of this class is created for each service replica by the Service Fabric runtime.
-	/// </summary>
+	/// </summary>B
```
(nit): Remove "B"

jessebenson commented on samples/Basic/Basic.ProductSvc/ProductSvc.cs:
```diff
@@ -51,11 +51,12 @@ protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListe
 
 					await products.SetAsync(tx, key, value, TimeSpan.FromSeconds(4), cancellationToken).ConfigureAwait(false);
 					await tx.CommitAsync().ConfigureAwait(false);
+                 //products.SetAsync()
```
(nit): delete

jessebenson commented on samples/Basic/Basic.WebSvc/Controllers/ProductsController.cs:
```diff
@@ -48,7 +48,7 @@ public class ProductsController : ApiController
 
 		private IProductService GetProductService(string sku)
-		{
+		{   //IMP888
```
(nit): delete?

jessebenson commented on samples/Basic/Basic.WebSvc/Controllers/QueryController.cs:
```diff
@@ -38,6 +40,144 @@ public Task<IHttpActionResult> GetMetadata(string application, string service)
+		public Task<IHttpActionResult> DeleteAsync(string application, string service, string collection, [FromBody] ValueViewModel[] obj)
+		{
+			return base.DeleteAsync(application, service, collection, obj);
+        }
```
(nit): reformat file (whitespace).

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,7 +74,98 @@ public abstract class QueryableController : ApiController
+                    //Fetch partition proxy.
+			        var proxy = await GetServiceProxyForPartitionAsync<IQueryableService>(serviceUri, obj[i].PartitionId)
+			            .ConfigureAwait(false);
+                    //Perform delete opration.
```
(sp): "operation"

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/ValueViewModel.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.ServiceFabric.Services.Queryable.Controller
+{
+    public class ValueViewModel
```
Can we just use Entity<JToken, JToken> instead of this class?  Delete this class.

jessebenson replied:
Or add a "class Entity {}" with these form of the JToken properties.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/IQueryableService.cs:
```diff
@@ -35,5 +36,33 @@ public interface IQueryableService : IService
+	    /// <param name="key">Entity key inside reliable collection.</param>
+	    /// <returns>Boolean signifying the succes/failure operation.</returns>
+        Task<bool> DeleteAsync(string collection, string key);
+        /// <summary>
```
(nit): add empty line after methods/before comments.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -12,12 +12,17 @@
 
+
+
+
```
(nit):  remove extra whitespace.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/QueryableServiceBase.cs:
```diff
@@ -28,7 +29,24 @@ Task<IEnumerable<string>> IQueryableService.QueryAsync(string collection, IEnume
+	    }
+
+
+    }
```
(nit): whitespace alignment.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -64,58 +69,183 @@ internal static class ReliableStateExtensions
+	                var key = JsonConvert.DeserializeObject(keyJson, keyType);
+	                var val = JsonConvert.DeserializeObject(valJson, valueType);
+
+	                var dictionaryType = typeof(IReliableDictionary<,>).MakeGenericType(keyType, valueType);
```
refactor this code - create a private static method to get the dictionary type from dictionary instance.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -64,58 +69,183 @@ internal static class ReliableStateExtensions
+
+	                var dictionaryType = typeof(IReliableDictionary<,>).MakeGenericType(keyType, valueType);
+
+	                //Task<bool> TryUpdateAsync(ITransaction tx, TKey key, TValue newValue, TValue comparisonValue);
```
delete commented out line.

Commit:
Reformatted

jdhuntington commented on samples/Basic/Basic.WebSvc/Controllers/QueryController.cs:
```diff
@@ -37,7 +38,149 @@ public Task<IHttpActionResult> GetMetadata(string application, string service)
+		/// <summary>
+		/// Deletes appropriate key & corresponding value from the given reliable collection in the queryable service.
+		/// SINGLE DELETE:
+		/// - DELETE /query/BasicApp/ProductSvc/products and in Body provide a Json:
```
Having a body is awkward for a `DELETE` request. Most server implementations seem to ignore the payload, though it is not forbidden by the spec. I think a better implementation would be to send `DELETE` to the URI of a specific resource (eg `DELETE /query/BasicApp/ProductSvc/products/{partitionkey}/{key}`)

tushardhadiwal replied:
Using InBody method is necessary as key can be anything(nested objects) and of any length since it is user defined. @jessebenson

jdhuntington commented on samples/Basic/Basic.WebSvc/Controllers/QueryController.cs:
```diff
@@ -37,7 +38,149 @@ public Task<IHttpActionResult> GetMetadata(string application, string service)
+		///	    }]
+		/// 
+		/// Update a Key with New Value in all Partitions:
+		/// -PUT /query/BasicApp/ProductSvc/products 
```
I don't think update across partitions is something users will expect, and I think we'd be better off omitting it.

jdhuntington commented on samples/Basic/Basic.WebSvc/Controllers/QueryController.cs:
```diff
@@ -37,7 +38,149 @@ public Task<IHttpActionResult> GetMetadata(string application, string service)
+		/// 
+		/// 
+		/// BATCH UPDATE (Provide an Array of keys & values with optional partitionID in JSON format inside body of HTTP PUT request to update all the keys with new values.). 
+		/// -PUT /query/BasicApp/ProductSvc/products
```
I'd like to see a URI for each entity, rather than having it partially encoded in the URL and partially in the body. `PUT /query/BasicApp/ProductSvc/products/39704134134861/sku-218` feels more user-friendly.

tushardhadiwal replied:
Same reason as above. @jessebenson

jdhuntington commented on samples/Basic/Basic.WebSvc/Controllers/QueryController.cs:
```diff
@@ -37,7 +38,149 @@ public Task<IHttpActionResult> GetMetadata(string application, string service)
+			return base.DeleteAsync(application, service, collection, obj);
+		}
+
+		/// <summary>
```
The docs/comments should include a description of what happens if the entity already exists.

tushardhadiwal replied:
Added!

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -183,30 +322,42 @@ private static Type GetValueType(this IReliableState state)
 				var asyncEnumerator = asyncEnumerable.CallMethod<object>("GetAsyncEnumerator");
 
 				// Copy all items from the reliable dictionary into memory.
+				// TODO: cache the method/property objects and invoke with new parameters
```
Remove TODO

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -56,66 +60,192 @@ internal static class ReliableStateExtensions
+			}
+			catch (ArgumentException)
+			{
+
```
nit: spacing

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -12,9 +12,13 @@
 using System.Xml;
+using Microsoft.ServiceFabric.Services.Queryable.Controller;
+
+
```
nit: spacing

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,15 +75,113 @@ public abstract class QueryableController : ApiController
+						.ConfigureAwait(false);
+
+				}
+				return Ok(results);
```
This should return something more descriptive. A mapping of keys to successes would be helpful. Alternately, I'm not sure this needs to delete more than 1 thing at a time.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/QueryableService.cs:
```diff
@@ -25,11 +26,28 @@ Task<string> IQueryableService.GetMetadataAsync()
 		}
-	}
+
+
```
nit: Inconsistent spacing in this file.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -56,66 +60,192 @@ internal static class ReliableStateExtensions
-				var objects = results.Select(r => JsonConvert.DeserializeObject(r, valueType));
-				var queryResult = ApplyQuery(objects, valueType, query, aggregate: true);
+				var entityType = reliableState.GetEntityType();
+				// var temp= results.Select(x=>JsonConvert.DeserializeObject(x,))
```
Nit: Remove commented code.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,15 +75,113 @@ public abstract class QueryableController : ApiController
+					results[i] = await proxy.AddAsync(collection, keyquoted, valuequoted).ConfigureAwait(false);
+
+				}
+				return Ok(results);
```
The result should include URIs of the entities created.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,15 +75,113 @@ public abstract class QueryableController : ApiController
+					var proxy = await GetServiceProxyForAddAsync<IQueryableService>(serviceUri, obj[i].PartitionId)
+						.ConfigureAwait(false);
+
+					results[i] = await proxy.AddAsync(collection, keyquoted, valuequoted).ConfigureAwait(false);
```
Instead of a for loop, you can map over the input creating new tasks, then await all of the tasks at the same time letting them run in parallel.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -95,12 +198,41 @@ private IHttpActionResult HandleException(Exception e, Uri serviceUri)
+				var matchingPartitions = partitions.Where(p => p.PartitionInformation.Id == partitionId);
+				if (partitionId == Guid.Empty)
+				{
+					return CreateServiceProxy<T>(serviceUri, partitions.First());
```
This doesn't seem like the optimal spread of partitions. Can you pick one at random? Do you need the user of the library to define a random partition method for you to call?

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Query/QueryModelCache.cs:
```diff
@@ -16,6 +16,8 @@ public ODataQueryContext GetQueryContext(Type type)
 		private static ODataQueryContext CreateQueryContext(Type type)
 		{
+
+            //here we build schema 
```
Spacing.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,15 +75,113 @@ public abstract class QueryableController : ApiController
+					results[i] = await Task.WhenAll(proxy.Select(p => p.UpdateAsync(collection, keyquoted, valuequoted)))
+						.ConfigureAwait(false);
+				}
+				return Ok(results);
```
Mapping update results to individual actions seems like a better UI, rather than a list of booleans.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,15 +75,113 @@ public abstract class QueryableController : ApiController
+				for (int i = 0; i < obj.Length; i++)
+				{
+					//Serialize the key from the json body and put it into a string.
+					string keyquoted = JsonConvert.SerializeObject(obj[i].Key,
```
It looks like these are being instantly deserialized by the callee on the other end. What is the value of encoding them?

Commit:
Codde review issues partially fixed.

Commit:
Delete Operation now groups keys by partitionID & deletes all keys belonging to one partition in one call.
Also, Sucess result is shown per PID & KEY in verbose manner.

Commit:
Inprogress

Commit:
IP

Commit:
Code Review formatting issues addressed.

Commit:
Code Review Fixes

Commit:
Codemaid reformatting.

Commit:
.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,6 +74,118 @@ public abstract class QueryableController : ApiController
+				for (int i = 0; i < obj.Length; i++)
+				{
+					//Serialize the key from the json body and put it into a string.
+					string keyquoted = JsonConvert.SerializeObject(obj[i].Key,
```
Looks like the comment got dropped on cleanup, but it's still unclear why we are serializing the keys only to deserialize them in the backend. Can this be removed?

jessebenson replied:
We hit issues where we couldn't send JToken over ServiceProxy - it was hitting exceptions trying to serialize it :(  - I agree with the point though, we should try to be more efficient here.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -79,28 +195,78 @@ private IHttpActionResult HandleException(Exception e, Uri serviceUri)
+				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
+				var matchingPartitions = partitions.Where(p => p.PartitionInformation.Id == partitionId);
+
+				int randomindex = random.Value.Next(0, partitions.Count);
```
~~If memory serves, the partition key is better if it's between INTMIN and INTMAX, though perhaps we don't actually have any guarantees about the distribution. @jessebenson Can you comment?~~

I don't think this is necessary - never mind.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -79,28 +195,78 @@ private IHttpActionResult HandleException(Exception e, Uri serviceUri)
+			using (var client = new FabricClient())
+			{
+				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
+				var matchingPartitions = partitions.Where(p => p.PartitionInformation.Id == partitionId);
```
You can replace `Where` with `First` and drop the `First` call 5 lines later.

tushardhadiwal replied:
Done.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,6 +74,118 @@ public abstract class QueryableController : ApiController
+
+					var proxy = await GetServiceProxyForPartitionAsync<IQueryableService>(serviceUri, obj[i].PartitionId)
+						.ConfigureAwait(false);
+					results[i] = await Task.WhenAll(proxy.Select(p => p.UpdateAsync(collection, keyquoted, valuequoted)))
```
Can we require partition-id here?  The same way we do for Delete?  If they're updating an entry, they should tell us exactly which one (and only one) they want updated.  Since the "same" key in two different partitions most likely has different meanings - we should not assume we should update them both.

tushardhadiwal replied:
Done.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,6 +74,118 @@ public abstract class QueryableController : ApiController
+			try
+			{
+				bool[] results = new bool[obj.Length];
+				Dictionary<JToken, bool> results1 = new Dictionary<JToken, bool>();
```
This could use a more descriptive name. Do we need to result sets?

tushardhadiwal replied:
I have changed entire Add to be more efficient just like Update & Delete & removed 2 result sets.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -70,6 +74,118 @@ public abstract class QueryableController : ApiController
+					results[i] = await proxy.AddAsync(collection, keyquoted, valuequoted).ConfigureAwait(false);
+					results1[obj[i].Key] = results[i];
+				}
+				return Ok(results1);
```
How does the caller know what partition ID a new entity was given?

jessebenson replied:
+1 - imo it would be nicer to see the return value as something like:

[{
      "PartitionId": 123456-0000-0000-0000,
      "Key": "key A",
      "Status":  200
   },
   {
      "PartitionId": 456789-0000-0000-0000,
      "Key": "key B",
      "Status":  200
   }]

jessebenson replied:
For add, update, and delete.

tushardhadiwal replied:
Done. Result is shown just like @jessebenson mentioned, along with that, we also group the keys by partitionID.

jdhuntington commented on src/Microsoft.ServiceFabric.Services.Queryable/Service/ReliableStateExtensions.cs:
```diff
@@ -56,66 +56,180 @@ internal static class ReliableStateExtensions
-				var objects = results.Select(r => JsonConvert.DeserializeObject(r, valueType));
-				var queryResult = ApplyQuery(objects, valueType, query, aggregate: true);
+				var entityType = reliableState.GetEntityType();
+				var objects = results.Select(r => JsonConvert.DeserializeObject(r, entityType));
```
Why are we invoking JSON?

jessebenson replied:
The code above queries all the other partitions (and gets back a bunch of JSON results).
We deserialize the JSON into objects, re-apply (some of) the query to sort/project/etc.
Then we serialize to JSON, because ServiceProxy serializes things via DataContract and throws an exception if you give it generic objects.

I think we will move to just HTTP on the stateful service side, which will help a bit going forward.

Commit:
All code review suggestions fixed.
Add is much more efficient now.
RandomPartitionID is generated & notified to user where record is put.
Unified PartitionProxy Creation for all.

Commit:
Changed the results for add delete and update to show per key without grouping by PID.
Added support for continual operation even if wrong data is provided in middle.
Changed the successBools to HTTPStatus Codes.

Commit:
Integration with SFX Begins.
Added HTTPOptions Support.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -83,7 +83,7 @@ public abstract class QueryableController : ApiController
 				Dictionary<Guid, List<JToken>> preMap = new Dictionary<Guid, List<JToken>>();
 
 				Dictionary<Guid, Dictionary<JToken, bool>> parResult = new Dictionary<Guid, Dictionary<JToken, bool>>();
```
Delete parResult line.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -134,6 +136,8 @@ public abstract class QueryableController : ApiController
 
 				Dictionary<Guid, Dictionary<JToken, bool>> parResult = new Dictionary<Guid, Dictionary<JToken, bool>>();
```
Delete 'parResult' line - it's unused.

jessebenson commented on src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs:
```diff
@@ -53,29 +59,49 @@ public abstract class QueryableController : ApiController
+			if (request.Headers.TryGetValues("Origin", out headers))
+				response.Headers.Add("Access-Control-Allow-Origin", headers);
+			if (request.Headers.TryGetValues("Access-Control-Request-Headers", out headers))
+				response.Headers.Add("Access-Control-Allow-Headers", headers);
```
These classes/APIs need to be updated for ASP.NET Core.

Commit:
This method is ported to ASP MVC in tushar branch

Commit:
ParResult removed.
